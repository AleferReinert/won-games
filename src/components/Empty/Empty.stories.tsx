import type { StoryObj, Meta } from '@storybook/react'
import { within } from '@storybook/testing-library'
import { expect } from '@storybook/jest'
import Empty from './Empty'

const meta: Meta<typeof Empty> = {
  title: 'Components/Empty',
  component: Empty,
  args: {
    title: 'No results found',
    description: `Sorry, we couldn't find any results for your search.`
  },
  argTypes: {
    link: { if: { arg: 'label' } }
  }
}

export default meta

type Story = StoryObj<typeof Empty>

export const Default: Story = {
  play: ({ canvasElement }) => {
    const canvas = within(canvasElement)
    const img = canvas.getByRole('img')
    const title = canvas.getByRole('heading', { name: /no results found/i })
    const description = canvas.getByText(
      /sorry, we couldn't find any results for your search./i
    )

    expect(img).toBeInTheDocument()
    expect(title).toBeInTheDocument()
    expect(description).toBeInTheDocument()
  }
}

export const WithButton: Story = {
  args: {
    label: 'Go back to store',
    link: '/link'
  },
  play: ({ canvasElement }) => {
    const canvas = within(canvasElement)
    const button = canvas.getByRole('link', { name: /go back to store/i })

    expect(button).toHaveAttribute('href', '/link')
  }
}
