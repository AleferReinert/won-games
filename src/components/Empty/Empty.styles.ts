import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    margin: ${theme.spacings.medium};
    text-align: center;
  `}
`

export const Img = styled.img`
  ${({ theme }) => css`
    max-width: 34rem;
    width: 100%;
    margin-bottom: ${theme.spacings.small};
    margin-right: -0.3rem;
  `}
`

export const Title = styled.h2`
  ${({ theme }) => css`
    color: ${theme.colors.primary};
    font-size: ${theme.font.sizes.xlarge};

    @media (min-width: ${theme.breakpoint.medium}) {
      font-size: ${theme.font.sizes.xxlarge};
    }
  `}
`

export const Message = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.light};
    margin-bottom: ${theme.spacings.small};

    @media (min-width: ${theme.breakpoint.medium}) {
      font-size: ${theme.font.sizes.medium};
    }
  `}
`
