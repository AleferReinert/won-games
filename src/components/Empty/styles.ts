import Image from 'next/image'
import styled, { css } from 'styled-components'

export const Wrapper = styled.section`
  ${({ theme }) => css`
    padding: ${theme.spacings.medium};
    text-align: center;
  `}
`

export const Img = styled(Image)`
  max-width: 38rem;
  width: 100%;
`

export const Title = styled.h2`
  ${({ theme }) => css`
    color: ${theme.colors.primary};
    font-size: ${theme.font.sizes.xlarge};

    ${theme.media().greaterThan('medium')`
        font-size: ${theme.font.sizes.xxlarge};
    `}
  `}
`

export const Message = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.light};
    margin-bottom: ${theme.spacings.xsmall};

    ${theme.media().greaterThan('medium')`
        font-size: ${theme.font.sizes.medium};
    `}
  `}
`
