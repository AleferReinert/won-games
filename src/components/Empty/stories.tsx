import { ComponentStory, ComponentMeta } from '@storybook/react'
import Empty from '.'

export default {
  title: 'Empty',
  component: Empty,
  args: {
    title: 'No results found',
    description: `Sorry, we couldn't find any results for your search.`,
    link: '/',
    label: 'Go back to store'
  }
} as ComponentMeta<typeof Empty>

export const Default: ComponentStory<typeof Empty> = (args) => (
  <Empty {...args} />
)
