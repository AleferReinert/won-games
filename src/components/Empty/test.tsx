import { screen } from '@testing-library/react'
import Empty, { EmptyProps } from '.'
import { renderWithTheme } from 'utils/tests/helpers'

const props: EmptyProps = {
  title: 'my title',
  description: 'my description'
}

describe('<Empty />', () => {
  it('should render img, title and description without link', () => {
    const { container } = renderWithTheme(<Empty {...props} />)

    expect(
      screen.getByRole('img', {
        name: /a person on a couch playing video games/i
      })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: props.title })
    ).toBeInTheDocument()
    expect(screen.getByText(props.description)).toBeInTheDocument()
    expect(
      screen.queryByRole('link', {
        name: /go back to store/i
      })
    ).not.toBeInTheDocument()

    expect(container.parentElement).toMatchSnapshot()
  })

  it('should render with link (label required)', () => {
    renderWithTheme(<Empty {...props} link='/' label='Go back to store' />)

    expect(
      screen.getByRole('link', {
        name: /go back to store/i
      })
    ).toHaveAttribute('href', props.link)
  })
})
