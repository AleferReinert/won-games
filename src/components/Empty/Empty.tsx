import Button from 'components/Button/Button'
import * as S from './Empty.styles'

export interface EmptyProps {
  title: string
  $description: string
  $invertedColors?: boolean
  $small?: boolean
  buttonText?: string
  buttonUrl?: string
}

const Empty = ({ title, $description, buttonText, buttonUrl, $invertedColors = false, $small = false }: EmptyProps) => {
  return (
    <S.Wrapper
      data-testid='EmptyComponent'
      title={title}
      $description={$description}
      $invertedColors={$invertedColors}
      $small={$small}
    >
      <S.Img
        src='/img/empty.svg'
        alt='A person on a couch playing video games'
        width={$small ? 140 : 340}
        height={$small ? 72 : 176}
      />
      <S.Title>{title}</S.Title>
      <S.Message>{$description}</S.Message>

      {buttonUrl && (
        <Button asLink href={buttonUrl}>
          {buttonText}
        </Button>
      )}
    </S.Wrapper>
  )
}

export default Empty
