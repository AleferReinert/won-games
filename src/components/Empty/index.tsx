import Button from 'components/Button'
import * as S from './styles'

export type EmptyPropsWithoutLink = {
  title: string
  description: string
  link?: undefined
  label?: undefined
}

export type EmptyPropsWithLink = {
  title: string
  description: string
  link: string
  label: string
}

export type EmptyProps = EmptyPropsWithLink | EmptyPropsWithoutLink

const Empty = ({ title, description, link, label }: EmptyProps) => {
  return (
    <S.Wrapper>
      <S.Img
        src='/img/empty.svg'
        fill
        alt='A person on a couch playing video games'
      />
      <S.Title>{title}</S.Title>
      <S.Message>{description}</S.Message>
      {!!link && (
        <Button as='a' href={link}>
          {label}
        </Button>
      )}
    </S.Wrapper>
  )
}

export default Empty
