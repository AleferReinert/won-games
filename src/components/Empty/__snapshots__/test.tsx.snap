// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Empty /> should render img, title and description without link 1`] = `
.c0 {
  margin: 3.2rem;
  text-align: center;
}

.c1 {
  max-width: 34rem;
  width: 100%;
  margin-bottom: 2.4rem;
  margin-right: -0.3rem;
}

.c2 {
  color: #F231A5;
  font-size: 2.0rem;
}

.c3 {
  color: #FAFAFA;
  font-size: 1.4rem;
  font-weight: 300;
  margin-bottom: 2.4rem;
}

@media (min-width:960px) {
  .c2 {
    font-size: 2.8rem;
  }
}

@media (min-width:960px) {
  .c3 {
    font-size: 1.6rem;
  }
}

<body>
  <div>
    <div
      class="c0"
    >
      <img
        alt="A person on a couch playing video games"
        class="c1"
        src="/img/empty.svg"
      />
      <h2
        class="c2"
      >
        my title
      </h2>
      <p
        class="c3"
      >
        my description
      </p>
    </div>
  </div>
</body>
`;
